Support "stay open" mode for Lambda invocations (updated PR)

whummer opened the pull request:
Support "stay open" mode for Lambda invocations.

Based on #4906 . Recreated the PR, to trigger the CI builds, and added a few minor fixes.

Replaces #4185
Closes #4123

whummer commented on localstack/services/awslambda/lambda_executors.py:
```diff
@@ -783,14 +794,53 @@ def execute_in_container(
-            stdin=stdin,
+        lambda_docker_ip = DOCKER_CLIENT.get_container_ip(container_info.name)
+
+        if not lambda_docker_ip or not in_docker():
```
Note: Over time we could add an additional check here that verifies whether the target endpoint (`f"http://{lambda_docker_ip}:9001"`) is actually accessible in `invoke_lambda(..)` below, and otherwise also fall back to `exec_in_container(..)`. 

This could help deal with issues where potentially the Lambda container is not accessible (e.g., because it's running in a different Docker network). Should usually not be the case (because Lambdas typically also call back to the LocalStack APIs on the same network), but technically possible..

whummer commented on localstack/services/awslambda/lambda_executors.py:
```diff
@@ -1410,18 +1477,6 @@ def docker_image_for_lambda(cls, lambda_function: LambdaFunction):
             )
         docker_tag = runtime
         docker_image = config.LAMBDA_CONTAINER_REGISTRY
-        # TODO: remove prefix once execution issues are fixed with dotnetcore/python lambdas
```
These patches seem to be no longer required - in fact, some of these older versions do not support `$DOCKER_LAMBDA_STAY_OPEN`, hence removing this code from here.

dfangl commented on localstack/services/awslambda/lambda_executors.py:
```diff
@@ -854,8 +904,13 @@ def prime_docker_container(
                 DOCKER_CLIENT.start_container(container_name)
+
+                def wait_up():
+                    cont_status = DOCKER_CLIENT.get_container_status(container_name)
```
We wait here for the container to start up, but would that necessarily include that the 9001 socket is already available? It only checks if the container status is up, not whether the port is already open. Might be a small window we never hit though.
This would also tie in nicely with the availability check for port 9001 to check against networking issues.

whummer replied:
Great point, thanks @dfangl . 👍  Updated the PR - it now also contains logic to assert availability of the target port, where appropriate (only in case we're running in Docker, and if the IP address of the target container can be determined by the Docker client).

whummer commented:
Thanks a bunch @chaz-doyle-cko . Rebased the branch and squashed the commits into two parts, to retain the history and authorship. 👍 

We'll also acknowledge the authorship of @jamietanna (original author) and @wesselvdv (contributor) in the commit messages. (will get merged once the builds are passing..)

Thanks all for contributing to this!

jamietanna commented:
Amazing, thanks @chaz-doyle-cko for getting this over the line, I know a lot of folks are going to really appreciate this :raised_hands:

Did we want to add anything extra in the docs to note this is better for Java Lambda performance?

wesselvdv commented:
When can we expect a release to be cut that includes this wonderful change?

jamietanna commented:
This should be available in the `:latest` tag if you want to give it a go in the meantime, btw :smile:

wesselvdv commented:
> This should be available in the `:latest` tag if you want to give it a go in the meantime, btw 😄

ohhhh, I should've thought of this myself. yay!
